### ISSUE TITLE
[Ray: data] Dataset.map_batches

### ISSUE THREAD

**Issue by suexu1025** [ray-project/ray#34068]
### Description

we need Dataset.map_batches to support output data type of TORCH.TENSOR 

### Use case

data transformation for Pytorch training input pipeline

**Comment by jianoaix** [comment]
@suexu1025 There is a `iter_torch_tensor()` API that produces torch tensors for you at the trainer. Could that work for your use case?

**Comment by suexu1025** [comment]
how to apply Dataset.map_batches before iter_torch_tensor()?
example code:

    def pd_to_tensor(batch: pd.DataFrame) -> torch.tensor:
        tensor  = torch.as_tensor(batch["images"], dtype=torch.float)
        # (B, H, W, C) -> (B, C, H, W)
        #tensor = tensor.permute(0, 3, 1, 2).contiguous()
        tensor = tensor.unsqueeze(0)
        # # [0., 255.] -> [0., 1.]
        tensor = tensor.div(255)
        return batch["images"], batch["label"]

ds = ds.map_batches(pd_to_tensor)

for step, data in enumerate(ds. iter_torch_tensor(batch_size=128)):
     pass

Errors:
it will generate output data type of TORCH.TENSOR is not supported.

**Comment by suexu1025** [comment]
detailed of error log:

returned a value of type <class 'torch.Tensor'>. This isn't allowed -- `map_batches` expects `fn` to return a `pandas.DataFrame`, `pyarrow.Table`, `numpy.ndarray`, `list`, or `dict[str, numpy.ndarray]`.

**Comment by raulchen** [comment]
You can return a `Dict[str, Tensor]` in `pd_to_tensor`.
